kotlinx-coroutines-core/jvm/test/linearizability/ChannelsLCStressTest.kt

```diff
@@ -0,0 +1,155 @@
+/*
+ * Copyright 2016-2019 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
+ */
+@file:Suppress("unused")
+
+package kotlinx.coroutines.linearizability
+
+import kotlinx.coroutines.*
+import kotlinx.coroutines.channels.*
+import kotlinx.coroutines.channels.TestChannelKind.*
+import kotlinx.coroutines.selects.*
+import org.jetbrains.kotlinx.lincheck.*
+import org.jetbrains.kotlinx.lincheck.annotations.*
+import org.jetbrains.kotlinx.lincheck.annotations.Operation
+import org.jetbrains.kotlinx.lincheck.execution.*
+import org.jetbrains.kotlinx.lincheck.paramgen.*
+import org.jetbrains.kotlinx.lincheck.strategy.stress.*
+import org.jetbrains.kotlinx.lincheck.verifier.*
+import org.junit.*
+import org.junit.runner.*
+import org.junit.runners.*
+import kotlin.reflect.jvm.*
+
+@Param.Params(
+    Param(name = "value", gen = IntGen::class, conf = "1:5"),
+    Param(name = "closeToken", gen = IntGen::class, conf = "1:3")
+)
+class ChannelsLCStressTestImpl : VerifierState() {
+    private val c = channelKind.create()
+
+    @Operation
+    suspend fun send(@Param(name = "value") value: Int) = try {
+        c.send(value)
+    } catch (e: NumberedCancellationException) {
+        e.testResult
+    }
+
+    @Operation
+    fun offer(@Param(name = "value") value: Int) = try {
+        c.offer(value)
+    } catch (e: NumberedCancellationException) {
+        e.testResult
+    }
+
+    @Operation
+    suspend fun sendViaSelect(@Param(name = "value") value: Int) = try {
+        select<Unit> { c.onSend(value) {} }
+    } catch (e: NumberedCancellationException) {
+        e.testResult
+    }
+
+    @Operation
+    suspend fun receive() = try {
+        c.receive()
+    } catch (e: NumberedCancellationException) {
+        e.testResult
+    }
+
+    @Operation
+    fun poll() = try {
+        c.poll()
+    } catch (e: NumberedCancellationException) {
+        e.testResult
+    }
+
+    @Operation
+    suspend fun receiveViaSelect() = try {
+        select<Int> { c.onReceive { it } }
+    } catch (e: NumberedCancellationException) {
+        e.testResult
+    }
+
+    @Operation
+    fun close(@Param(name = "closeToken") token: Int) = c.close(NumberedCancellationException(token))
+
+    // TODO: this operation should be (and can be!) linearizable, but is not
+    // @Operation
+    fun cancel(@Param(name = "closeToken") token: Int) = c.cancel(NumberedCancellationException(token))
+
+    @Operation
+    fun isClosedForReceive() = c.isClosedForReceive
+
+    @Operation
+    fun isClosedForSend() = c.isClosedForSend
+
+    // TODO: this operation should be (and can be!) linearizable, but is not
+    // @Operation
+    fun isEmpty() = c.isEmpty
```

Comment:
Can we still test `isEmpty` for rendezvous channels where it should be linearizable? It is not a short-stopper if we cannot, but would be a nice thing to have.

Reply:
Technically: yes, we can. For example, we can either always return `false` for other channels or create a special test for rendezvous one. However, it makes the code uglier. I would NOT test `isEmpty` since we have never had a test for it, and everything will be fixed with new channels.